# Data Cleaning

### Imports

In [1]:
import pandas as pd
import numpy as np

### Import CSV

In [2]:
df = pd.read_csv('data/pluto_19v1.csv')
df.head(5)

/opt/anaconda3/envs/capstone_env/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3145: DtypeWarning: Columns (16,17,18,20,22) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


borough  block  lot     cd  ct2010  cb2010  schooldist  council  zipcode  \
0      MN   1103   36  104.0   135.0  1003.0         2.0      6.0  10019.0   
1      MN   1109  100  104.0   135.0  1000.0         2.0      3.0  10019.0   
2      MN   1121   29  107.0   153.0  6000.0         3.0      6.0  10023.0   
3      MN   1099   12  104.0   135.0  2005.0         2.0      3.0  10019.0   
4      BK   5751   49  310.0   212.0  3002.0        20.0     38.0  11219.0   

  firecomp  ...  pfirm15_flag    rpaddate    dcasdate  zoningdate  landmkdate  \
0     E040  ...           NaN  07/06/2019  11/15/2018  07/29/2019  07/05/2019   
1     E040  ...           1.0  07/06/2019  11/15/2018  07/29/2019  07/05/2019   
2     L035  ...           NaN  07/06/2019  11/15/2018  07/29/2019  07/05/2019   
3     E040  ...           NaN  07/06/2019  11/15/2018  07/29/2019  07/05/2019   
4     E284  ...           NaN  07/06/2019  11/15/2018  07/29/2019  07/05/2019   

   basempdate masdate polidate  edesigdate  \
0  07/26/2019     NaN      NaN  05/31/2019   
1  07/26/2019     NaN      NaN  05/31/2019   
2  07/26/2019     NaN      NaN  05/31/2019   
3  07/26/2019     NaN      NaN  05/31/2019   
4  07/26/2019     NaN      NaN  05/31/2019   

                                                geom  
0  0106000020E61000000100000001030000000100000007...  
1  0106000020E6100000010000000103000000010000000A...  
2  0106000020E61000000100000001030000000100000009...  
3  0106000020E61000000100000001030000000100000009...  
4  0106000020E6100000010000000103000000010000000A...  

[5 rows x 95 columns]

In [3]:
#looking at features, null values and datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 859459 entries, 0 to 859458
Data columns (total 95 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   borough               859459 non-null  object 
 1   block                 859459 non-null  int64  
 2   lot                   859459 non-null  int64  
 3   cd                    857972 non-null  float64
 4   ct2010                858882 non-null  float64
 5   cb2010                857972 non-null  float64
 6   schooldist            857254 non-null  float64
 7   council               857971 non-null  float64
 8   zipcode               838521 non-null  float64
 9   firecomp              857234 non-null  object 
 10  policeprct            857247 non-null  float64
 11  healtharea            857248 non-null  float64
 12  sanitboro             857052 non-null  float64
 13  sanitsub              856932 non-null  object 
 14  address               858986 non-null  object 
 15  

In [4]:
df.isna().sum()

borough            0
block              0
lot                0
cd              1487
ct2010           577
               ...  
basempdate         0
masdate       859459
polidate      859459
edesigdate         0
geom            1915
Length: 95, dtype: int64

In [5]:
df.ext.value_counts()

N     451333
G     264413
E      48335
EG     36709
Name: ext, dtype: int64

In [6]:
df.landuse.value_counts()

1.0     565356
2.0     131378
4.0      54508
11.0     27715
5.0      21482
3.0      12846
8.0      12177
6.0      10497
10.0     10224
7.0       6335
9.0       4017
Name: landuse, dtype: int64

In [7]:
# for land use category, filter out 01 one and two family dwellings, 09 Open Space & Outdoor Recreation, 11 Vacant Land
# the remainder is lots with buildings

df = df[df['landuse'] != (1.0 or 9.0 or 11.0)]

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 294103 entries, 0 to 859458
Data columns (total 95 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   borough               294103 non-null  object 
 1   block                 294103 non-null  int64  
 2   lot                   294103 non-null  int64  
 3   cd                    292642 non-null  float64
 4   ct2010                293529 non-null  float64
 5   cb2010                292642 non-null  float64
 6   schooldist            291927 non-null  float64
 7   council               292641 non-null  float64
 8   zipcode               273332 non-null  float64
 9   firecomp              291910 non-null  object 
 10  policeprct            291922 non-null  float64
 11  healtharea            291923 non-null  float64
 12  sanitboro             291728 non-null  float64
 13  sanitsub              291611 non-null  object 
 14  address               293630 non-null  object 
 15  

In [8]:
pd.set_option('max_rows', 1000)

In [9]:
df.isna().sum()

borough                      0
block                        0
lot                          0
cd                        1461
ct2010                     574
cb2010                    1461
schooldist                2176
council                   1462
zipcode                  20771
firecomp                  2193
policeprct                2181
healtharea                2180
sanitboro                 2375
sanitsub                  2492
address                    473
zonedist1                 2504
zonedist2               281718
zonedist3               293914
zonedist4               294090
overlay1                234299
overlay2                293945
spdist1                 259766
spdist2                 294026
spdist3                 294103
ltdheight               292234
splitzone                 2504
bldgclass                  473
landuse                   2924
easements                  473
ownertype               263580
ownername                 1565
lotarea                   1551
bldgarea

In [10]:
df['geom'].value_counts()

0106000020E6100000010000000103000000010000000500000066009A5BA37A52C029DBE73E5458444074A4F977A27A52C0F9867E0D565844408882119C9D7A52C0BBD6DC82505844402FB7B07F9E7A52C0A18744B44E58444066009A5BA37A52C029DBE73E54584440                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    1
0106000020E61000000100000001030000000100000018000000B6CF737C618A52C0834455821246444076DE4734618A52C086E039FE0F464440A47E12BE628A52C05B1DDE960F4644405B121E69628A52C0B0E72BAA0C

In [11]:
df['edesigdate'].value_counts()

05/31/2019    294103
Name: edesigdate, dtype: int64

In [12]:
df['dcasdate'].value_counts()

11/15/2018    294103
Name: dcasdate, dtype: int64

In [13]:
df['masdate'].value_counts()

Series([], Name: masdate, dtype: int64)

In [14]:
df['rpaddate'].value_counts()

07/06/2019    294103
Name: rpaddate, dtype: int64

In [15]:
df['zmcode'].value_counts()

Y    6783
Name: zmcode, dtype: int64

In [16]:
df['ownertype'].value_counts()

X    17229
C    11619
O     1082
P      519
M       74
Name: ownertype, dtype: int64

In [17]:
df['numbldgs'].value_counts() #this shows the number of buildings on each lot

1.0      218637
0.0       34009
2.0       33256
3.0        3551
4.0        1201
5.0         494
6.0         294
7.0         162
8.0         134
9.0         102
10.0        102
12.0         61
11.0         55
13.0         44
14.0         35
15.0         33
16.0         27
17.0         26
19.0         18
21.0         16
18.0         15
22.0         13
20.0         12
32.0         10
28.0         10
23.0         10
24.0          9
27.0          9
31.0          8
26.0          8
53.0          7
39.0          6
43.0          6
54.0          6
25.0          5
41.0          5
35.0          5
100.0         5
33.0          5
44.0          4
52.0          4
78.0          4
101.0         4
29.0          4
46.0          4
34.0          3
42.0          3
61.0          3
37.0          3
30.0          3
92.0          3
90.0          3
38.0          3
64.0          3
109.0         3
119.0         2
51.0          2
50.0          2
68.0          2
66.0          2
80.0          2
83.0          2
48.0    

In [18]:
# filter to show only the lots with one building
df = df[df['numbldgs'] == 1.0]

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 218637 entries, 1 to 859457
Data columns (total 95 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   borough               218637 non-null  object 
 1   block                 218637 non-null  int64  
 2   lot                   218637 non-null  int64  
 3   cd                    218494 non-null  float64
 4   ct2010                218618 non-null  float64
 5   cb2010                218494 non-null  float64
 6   schooldist            218422 non-null  float64
 7   council               218494 non-null  float64
 8   zipcode               217176 non-null  float64
 9   firecomp              218419 non-null  object 
 10  policeprct            218423 non-null  float64
 11  healtharea            218423 non-null  float64
 12  sanitboro             218402 non-null  float64
 13  sanitsub              218359 non-null  object 
 14  address               218637 non-null  object 
 15  

In [19]:
df['bldgclass'].value_counts()

C0    58290
C1    14355
C3    13477
S2    12888
C2    11987
K1     7149
C7     6826
K4     5116
S1     4932
D1     4709
S9     3399
E1     3393
K2     3266
D4     3120
M1     3105
RM     3053
C6     2854
C4     2738
S3     2604
S5     2353
C5     2342
V0     2298
G0     2263
S4     2206
D7     1886
G2     1737
R1     1541
G1     1466
O5     1435
E9     1394
F5     1312
O2     1241
G7     1111
V1     1090
O7     1045
Z9     1020
D3     1005
R3      917
D6      909
R4      859
F4      834
W1      812
O6      797
M9      785
R6      550
O8      522
K5      517
D9      490
F9      420
F1      410
M3      410
O1      406
W2      402
G9      402
K7      401
O4      383
W9      337
R2      329
I5      322
Q1      313
E2      304
L8      292
H3      274
D5      267
Q2      258
N2      254
D0      245
P5      242
G4      239
N9      231
I9      217
G8      204
Y1      200
E7      199
U7      195
P8      193
RC      187
S0      178
K6      169
W3      165
C9      161
I7      160
G6      155
L9  

__Comments__

The following columns will be dropped for too many missing values:
- masdate
- polidate

The following columns will be dropped beause they are not relevant.
- zmcode, a code which identifies a tax lot on the border of two zoning maps.  
- condono, the condominium number.
- owner name
- ct2010, the 2010 census tract in which the tax lot is located.
- tract2010, The 2010 census tract in which the tax lot is located. 
- cb2010, the 2010 census block in which the tax lot is located.
- firecomp, the fire company that services the tax lot.
- sanborn, the Sanborn Map Company map number associated with the tax block and lot.
- sanitsub, the subsection of the sanitation district that services the tax lot.
- address
- areasource, a code indicating the methodology used to determine the tax lot's TOTAL BUILDING FLOOR AREA (BldgArea) 
- zonemap, Department of City Planning Georeferenced NYC Zoning Maps 
- taxmap, the Department of Finance paper tax map volume number associated with the tax block and lot.
- appbbl, the originating BBL (borough, block and lot) from the apportionment prior to the merge, split or property’s conversion to a condominium.   
- appdate, the date of the apportionment to merge or split the property or convert to a condominium. 
- plutomapid, a code indicating whether the tax lot is in the PLUTO file, the MapPLUTO file with water areas included, and/or the MapPLUTO file that is clipped to the shoreline.
- rpaddate, not sure what this date represents, all values have the same date (not listed in the data dictionary.)
- dcasdate, not sure what this date represents, all values have the same date (not listed in the data dictionary.)
- edesigdate, not sure what this date represents, all values have the same date (not listed in the data dictionary.)
- geom, a code of numbers and letters that is illegible.
- zipcode, no need to sort by zip for the goal of the project.

The following columns will be dropped because they are redundant with another column.
- sanitboro, redundant with borough
- Borough, redundant with borocode
- version, the version number for this release of PLUTO.

The following rows with missing values will be dropped:
- yearbuilt

__Note:__
- this data set is by tax lot rather than by building. Some tax lots have more than one building. This will need to be filtered out.

In [20]:
df.drop(axis=1, columns=['masdate', 'polidate', 'zmcode', 'condono', 'ownername', 'ct2010', 'tract2010', 'cb2010', 'firecomp',
                 'sanborn', 'sanitsub', 'address', 'areasource', 'zonemap', 'taxmap', 'appbbl', 'appdate', 'plutomapid',
                 'rpaddate', 'dcasdate', 'edesigdate', 'geom', 'sanitboro', 'borough', 'version', 'zmcode', 'zipcode'], 
        inplace=True)

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 218637 entries, 1 to 859457
Data columns (total 69 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   block                 218637 non-null  int64  
 1   lot                   218637 non-null  int64  
 2   cd                    218494 non-null  float64
 3   schooldist            218422 non-null  float64
 4   council               218494 non-null  float64
 5   policeprct            218423 non-null  float64
 6   healtharea            218423 non-null  float64
 7   zonedist1             218411 non-null  object 
 8   zonedist2             8634 non-null    object 
 9   zonedist3             86 non-null      object 
 10  zonedist4             4 non-null       object 
 11  overlay1              49740 non-null   object 
 12  overlay2              113 non-null     object 
 13  spdist1               23403 non-null   object 
 14  spdist2               59 non-null      object 
 15  

In [21]:
# drop rows with nan in yearbuilt column
df.dropna(axis=0, how='any', inplace=True, subset=['yearbuilt'])
df.isna().sum()

block                        0
lot                          0
cd                         143
schooldist                 215
council                    143
policeprct                 214
healtharea                 214
zonedist1                  226
zonedist2               210003
zonedist3               218551
zonedist4               218633
overlay1                168897
overlay2                218524
spdist1                 195234
spdist2                 218578
spdist3                 218637
ltdheight               216881
splitzone                  226
bldgclass                    0
landuse                   1084
easements                    0
ownertype               204641
lotarea                      0
bldgarea                     3
comarea                   9732
resarea                   9732
officearea                9732
retailarea                9732
garagearea                9732
strgearea                 9732
factryarea                9732
otherarea                 9732
numbldgs

### Prep for merging dataframes

In [22]:
# look at bbl, the dataframes will be merged on this column.
df.iloc[5]

block                                                                1127
lot                                                                    29
cd                                                                    107
schooldist                                                              3
council                                                                 6
policeprct                                                             20
healtharea                                                           3500
zonedist1                                                            R10A
zonedist2                                                             R8B
zonedist3                                                             NaN
zonedist4                                                             NaN
overlay1                                                              NaN
overlay2                                                              NaN
spdist1                               

__Comments__\
This dataframe will be filtered using BBL number. \
BBL stands for borough, block and lot. \
example bbl number 1011270029:\
borough 1 for Manhattan\
block 01127\
lot 0029

In [23]:
#send cleaned file to csv
df.to_csv('data/nyc_lots.csv')

__Columns Used for Model__

block  - location identifier which may be used if concatenating with other data 

lot - location identifier which may be used if concatenating with other data 

bbl, a concatenation of the borough code, tax block and tax lot - keep in order to merge with demo permit data. 

cd - Community District, of the 59 community districts in NYC. Different districts may tend to maintain buildings better or worse, or have different development patterns.                
schooldist - school district, affects population movement patterns which drive density and development.

council - city council district. Different districts may tend to maintain buildings better or worse, or have different development patterns.               

policeprct - police precinct, relevant because crime affects property values and development patterns.

healtharea - Health areas were originally created in the 1920s for the purpose of reporting and  statistical analysis of public health data. They were based on census tracts and created to be areas of equal population. Health areas are contained within health center districts. Buildings affect public health and a


For the remaining columns after other cleaning procedures, refer to the 01-Cleaning-Permits file.